src/pybind/mgr/dashboard_v2/tools.py

```diff
@@ -521,11 +534,207 @@ def run(self):
+            'namespace': t.namespace,
+            'metadata': t.metadata,
+            'begin_time': t.begin_time,
+            'end_time': t.end_time,
+            'latency': t.latency,
+            'progress': t.progress,
+            'success': not t.exception,
+            'ret_value': t.ret_value,
+            'exception': t.exception
+        } for t in fn_t]
+
+
+class AsyncTask(object):
+    # pylint: disable=too-many-instance-attributes
+
+    class ExecutorThread(threading.Thread):
+        def __init__(self, task):
+            super(AsyncTask.ExecutorThread, self).__init__()
+            self._task = task
+            self.event = threading.Event()
+
+        # pylint: disable=broad-except
+        def run(self):
+            TaskManager._task_local_data.task = self._task
+            try:
+                self._task.set_progress(0)
+                self._task.begin_time = time.time()
+                val = self._task.fn(*self._task.fn_args, **self._task.fn_kwargs)
+                self._task.end_time = time.time()
+            except Exception as ex:
+                logger.exception("Error while calling %s: ex=%s", self._task, str(ex))
+                with self._task.lock:
+                    self._task.end_time = time.time()
+                    self._task.ret_value = None
+                    self._task.exception = ex
+            else:
+                with self._task.lock:
+                    self._task.latency = self._task.end_time - self._task.begin_time
+                    self._task.ret_value = val
+                    self._task.exception = None
+                    self._task.set_progress(100, True)
+                logger.debug("execution of %s finished in: %s s", self._task, self._task.latency)
+            finally:
+                NotificationQueue.new_notification('cd_task_finished', self._task)
+                self.event.set()
+
+    def __init__(self, namespace, metadata, fn, args, kwargs):
+        self.namespace = namespace
+        self.metadata = metadata
+        self.fn = fn
+        self.fn_args = args
+        self.fn_kwargs = kwargs
+        self.executor_thread = None
+        self.event = threading.Event()
+        self.progress = None
+        self.ret_value = None
+        self.begin_time = None
+        self.end_time = None
+        self.latency = 0
+        self.exception = None
+        self.lock = threading.Lock()
+
+    def __hash__(self):
+        return hash((self.namespace, tuple(sorted(self.metadata.items()))))
+
+    def __eq__(self, other):
+        return self.namespace == self.namespace and self.metadata == self.metadata
+
+    def __str__(self):
+        return "Task(ns={}, md={})" \
+               .format(self.namespace, self.metadata)
+
+    def _run(self):
+        with self.lock:
+            if self.executor_thread is None:
+                self.executor_thread = AsyncTask.ExecutorThread(self)
+                self.executor_thread.start()
```

Comment:
You're unconditionally creating a new posix thread per scheduled task here and dashboard_v2 already starts about ten posix threads.

Reply:
But wouldn't each task have to be handled in a separate thread? How else would you execute them in parallel?

Reply:
@LenzGr I think what @sebastian-philipp means is that with the current implementation the number of threads might grow unbounded. I agree that we can improve this part of code by using a thread queue that limits the number of worker threads. I think this improvement can be done in a separate PR and is not critical at this point.